Pin versions of external docker images for more stability

ptrovatelli opened the pull request:
We have discussed it a little bit in slack: this PR pins the external docker image versions in order to have a reproducible build and avoid issues like the one we had recently on mysql change within the python image.
I have also pinned the digests because identifiers like python:3.7.4-slim-buster are not immutable, meaning that the image can change and keep the same identifier. 
After this we'll need to update versions from time to time. 
I have found a project that automatically creates pull request when a new image is released we could try that? It's free for open source projects:
https://renovatebot.com/blog/docker-mutable-tags
https://github.com/marketplace/renovate/plan/MDIyOk1hcmtldHBsYWNlTGlzdGluZ1BsYW45NTk=#pricing-and-setup

NB: i have found that image id is different from digest. 
What needs to be used in "FROM" and "docker pull" is the digest, not the image id
To find an image digest:
```
$docker pull python:3.7.4-buster
3.7.4-buster: Pulling from library/python
Digest: sha256:7dd57761929672e7c2c18348d257ed5148fc6a971a13cfd8a3d1d8573226a5c2
Status: Image is up to date for python:3.7.4-buster
```
This digest can be used in dockerfile "from"
However the image id cannot: 
```
$docker images --no-trunc
REPOSITORY                     TAG                 IMAGE ID                                                                  CREATED             SIZE
python                         3.7.4-slim-buster   sha256:53951a775ee29fec2ff8f56b941c0a445b7569b7e04daae3861616000988ad1a   9 hours ago         179MB
```

This also shows image id and digests: 
https://github.com/docker-library/repo-info/blob/master/repos/python/tag-details.md

madchap commented:
Should not we pin the image against `python:3.5.7-slim-buster` sha256 instead, so that our running docker image is in line with what we can do with travis and all? (As hinted by @Maffooch at #1318)

ptrovatelli commented:
@madchap OK. I was trying to set up the current equivalent (3.7.4 is the current latest) but if we agree on using python 3.5 I can pin this.

ptrovatelli commented:
@madchap Done. 
I have also added a link from README.MD to CONTRIBUTING.MD ; just discovered this file by the way! 
I have added a few lines about logging deduplication events (been looking for a file like this for info about devevelopments :) )

madchap commented:
Well, by no means am I a maintainer of this project, so I think final word would go to them. However, this would sound fairly logical there. Cheers!

Maffooch commented:
@madchap is correct. The image should be pinned against 3.5.x as that is the highest supported version that Travis is running on. It would be too much work (and too experimental for my taste) to get everything up tp python 3.7.x. I really like that renovate plugin idea. I believe we should be able to take advantage of it for free as well. I'm good with everything here. However, I would like other @DefectDojo/defectdojo-moderators  to weigh in here. I'm sure there is is some angle I'm not considering here. Great work on the changes and research.

ptrovatelli commented:
> @ptrovatelli Could you also look into what it would take to get the renovate bot included in our project?

@Maffooch yeahp i'll look into it

ptrovatelli commented:
@aaronweaver @Maffooch  so i've started looking at this renovate bot thing :) 
I don't have the permission to install it on the official DD repo (needs ownership i think, it's a github application) however I have installed it on my fork so you can have a look at it here: 
https://github.com/ptrovatelli/django-DefectDojo/pull/1
The configuration of the bot can be done by pushing /renovate.json file onto the renovate/configure branch (this is the "onboarding" branch)
The bot automatically updates it's initial comment in the PR to describe what it's about to do
After we merge the PR, it will do the actions (creates mores PR) with a limit of 2 per hour with the current configuration. 
Right now I've only changed the configuration in order to analyse dev branch, not master. 
You can have a look, maybe create PR on my repo if you want to fiddle with the configuration before installing the bot on the main DD repository

More info: 
https://github.com/marketplace/renovate
https://renovatebot.com/docs/
https://renovatebot.com/docs/configure-renovate
https://renovatebot.com/docs/reconfigure-renovate/
https://renovatebot.com/docs/configuration-options/
https://renovatebot.com/docs/faq/

Maffooch commented:
@ptrovatelli Thank you so much for taking the time to do this. I think It is super cool and should be implemented on both dev and master. Maybe two hours  per push is a bit too frequent though. A happy medium should be around twelve hours or so. Seems like it can be edited at any time as well. That's a big plus for me. What do you think @aaronweaver @devGregA

ptrovatelli commented:
@Maffooch it's maximum 2 PR created per hour. it seems a reasonable default. this is to avoid overflooding the CI. 
Right now, there are 26 planned PR, so it means it'll take 13h for the bot to create all the PR.
It think it's enough to analyse dev branch. it may be a hustle to make decisions and merge them both on master and dev.
Also master should remain stable and we may create issues by doing those merges without further tests

Maffooch commented:
Oh! I read that wrong. I guess that won't be too bad. For the master branch, we could do the additional testing when testing the PR. A new release is In the very near future as well. Maybe we could do a full run on the dev branch before the release. See if we like it and ensure everything is stable before adding it to the master branch.
